File: preprocess_lc_data.py
import os
import json
import numpy as np
import geopandas as gpd
import rioxarray
from rasterio.enums import Resampling

# Load config
with open("preprocess_config.json") as f:
    config = json.load(f)

# Paths from config
raster_path = config["land_cover_file"]
roi_path = config["roi_file"]
morph_folder = config["morph_folder"]
lc_folder = config["lc_folder"]
dem_file = os.path.join(morph_folder, "dem.asc")
target_resolution = config["cellsize"]
temp_dir = os.path.join(lc_folder, "temp")
os.makedirs(temp_dir, exist_ok=True)

# Reclassification dictionary
reclass_dict = {
    100: 2, 110: 2, 120: 2, 130: 2, 140: 2, 150: 2,
    200: 1, 210: 1, 211: 1, 212: 1, 220: 1, 221: 1, 222: 1,
    230: 1, 231: 1, 232: 1, 240: 1, 241: 1, 242: 1, 250: 1, 251: 1, 252: 1,
    300: 2, 310: 2, 311: 2, 312: 2, 320: 2, 330: 2,
    400: 2, 410: 2, 420: 1, 430: 2, 440: 2, 450: 2,
    500: 2, 510: 2, 520: 2, 530: 2,
    600: 2, 610: 2, 620: 2, 630: 2, 640: 2,
    800: 3,
    900: 2, 910: 3, 920: 2, 930: 3, 931: 3, 932: 3,
    1010: 3, 1020: 3,
    1200: 3
}

# Load ROI
roi = gpd.read_file(roi_path)

# Step 1: Load LC raster and clip to ROI
lc = rioxarray.open_rasterio(raster_path, masked=True).squeeze()
lc = lc.rio.write_crs("EPSG:4326", inplace=True) if lc.rio.crs is None else lc
lc_clipped = lc.rio.clip(roi.geometry.values, roi.crs, drop=True)

# Save clipped version
clipped_path = os.path.join(temp_dir, "landcover_clipped.tif")
lc_clipped.rio.to_raster(clipped_path)

# Step 2: Reclassify using dict
data = lc_clipped.data
reclassified = np.copy(data)
for old, new in reclass_dict.items():
    reclassified[data == old] = new
lc_clipped.data = reclassified

# Save reclassified version
reclass_path = os.path.join(temp_dir, "landcover_reclassified.tif")
lc_clipped.rio.to_raster(reclass_path)

# Step 3: Reproject to EPSG:4326 (safeguard, should already be EPSG:4326)
lc_clipped = lc_clipped.rio.reproject("EPSG:4326", resampling=Resampling.nearest)

# Step 4: Read DEM and align LC to its grid
dem = rioxarray.open_rasterio(dem_file, masked=True).squeeze()
lc_aligned = lc_clipped.rio.reproject_match(dem, resampling=Resampling.nearest)

# Step 5: Export final LC raster as ASCII Grid
final_lc_path = os.path.join(lc_folder, "landcover.asc")
lc_aligned.rio.to_raster(final_lc_path, driver="AAIGrid", nodata=-9999)

print(f"✅ Final landcover.asc saved to: {final_lc_path}")
print(f"🗂️ Clipped raster saved to: {clipped_path}")
print(f"🗂️ Reclassified raster saved to: {reclass_path}")

File: R/preprocess/preprocess_lc_data.R
# procesar DEM para mHM
{
  # HEADER --------------------------------------------
  #
  # Author: Diego Dinamarca
  # Email:  ddinamarcamuller@gmail.com
  # 
  # Date:
  #
  # Script Name: 
  #
  # Script Description: preparar datos de soilgrids para modelacion
  #
  # Notes:
  #
  #
  # INSTALL PACKAGES & LOAD LIBRARIES -----------------
  cat("INSTALLING PACKAGES & LOADING LIBRARIES... \n\n", sep = "")
  packages <- c("tidyverse",
                "terra", 
                "sf",
                "here",
                "jsonlite") # list of packages to load
  n_packages <- length(packages) # count how many packages are required
  
  new.pkg <- packages[!(packages %in% installed.packages())] # determine which packages aren't installed
  
  # install missing packages
  if(length(new.pkg)){
    install.packages(new.pkg)
  }
  
  # load all requried libraries
  for(n in 1:n_packages){
    cat("Loading Library #", n, " of ", n_packages, "... Currently Loading: ", packages[n], "\n", sep = "")
    lib_load <- paste("library(\"",packages[n],"\")", sep = "") # create string of text for loading each library
    eval(parse(text = lib_load)) # evaluate the string to load the library
  }
  # SET WORKING DIRECTORY -----------------------------
  cat("SETTING WORKING DIRECTORY...\n\n", sep = "")
  wd <- here::here()
  setwd(wd)
  cat("WORKING DIRECTORY HAS BEEN SET TO: ", wd, sep = "")
  
  # SET OPTIONS ---------------------------------------
  cat("SETTING OPTIONS... \n\n", sep = "")
  options(scipen = 999) # turns off scientific notation
  options(encoding = "UTF-8") # sets string encoding to UTF-8 instead of ANSI
  
  
  # CONFLICTS ---------------------------------------------------------------
  conflicted::conflict_prefer("select", "dplyr")
  conflicted::conflict_prefer("filter", "dplyr")
  conflicted::conflict_prefer("extract", "terra")
  conflicted::conflict_prefer("last", "dplyr")
  # LOAD FUNCTIONS ------------------------------------
  source("scripts/R/utils.R")
  
}

# leer archivo de configuracion json con rutas de archivos
config <- fromJSON("/Users/mhm/Desktop/FONDECYT_CAMILA/mhm_snow/SCRIPTS/preprocess_config.json")

# Paths from config
raster_path <- config$land_cover_file
roi_path <- config$roi_file
morph_folder <- config$morph_folder
lc_folder <- config$lc_folder
ref_file <- file.path(config$meteo_folder, "pre.nc")
cellsize <- config$cellsize

temp_dir <- file.path(lc_folder, "temp")
if (!dir.exists(temp_dir)) dir.create(temp_dir, recursive = TRUE)

# Reclassification dictionary
reclass_dict = as.matrix(rbind(
  c(100,150,2),
  c(200,252,1),
  c(300,410,2),
  c(411,429,1),
  c(430,640,2),
  c(790,810,3),
  c(890,910,2),
  c(909,911,3),
  c(912,929,2),
  c(930,1200,3)
))
reclass_dict
# Load ROI
roi <- st_read(roi_path, quiet = TRUE)
box = get_extent(roi)

# Crear buffer de 4km (4000 metros)
roi_buff <- st_buffer(box, dist = 4000)

# Step 1: Load LC raster and clip to ROI
lc <- rast(raster_path)
lc_clipped <- crop(lc, vect(st_transform(roi_buff, 32719)))

# Save clipped raster
clipped_path <- file.path(temp_dir, "landcover_clipped.tif")
writeRaster(lc_clipped, clipped_path, overwrite = TRUE)

# Step 2: Reclassify using dict
lc_reclass = classify(lc_clipped, rcl = reclass_dict, include.lowest = T, right = T)
lc_reclass[lc_reclass == 0] = NA
plot(lc_reclass)

# Save reclassified raster
reclass_path <- file.path(temp_dir, "landcover_reclassified.tif")
writeRaster(lc_reclass, reclass_path, overwrite = TRUE)

# Step 3: Reproject to EPSG:4326 (should already be)
lc_reproj <- project(lc_reclass, "EPSG:4326", method = "near")

# Step 4: Read DEM and align LC to its grid
ref <- rast(ref_file)[[1]]
res(ref) = cellsize
# roi_buff <- st_buffer(box.pol, dist = 3000)
# lc_3km= crop(lc_reproj, roi_buff)
lc_rs <- resample(lc_reproj, ref, method = "near")

# Step 5: Export final LC raster as ASCII Grid
# Crear buffer de 5km (3000 metros)
final_lc_path <- file.path(lc_folder, "landcover.asc")
writeRaster(lc_rs, final_lc_path, overwrite = TRUE, NAflag = -9999, datatype = "INT4S")

cat("✅ Final landcover.asc saved to:", final_lc_path, "\n")
